lib/qless/middleware/timeout.rb

```diff
@@ -0,0 +1,43 @@
+require 'timeout'
+
+module Qless
+  # Unique error class used when a job is timed out by this middleware.
+  # Allows us to differentiate this timeout from others caused by `::Timeout::Erorr`
+  JobTimedoutError = Class.new(StandardError)
+
+  module Middleware
+    # Applies a hard time out. To use this middleware, instantiate it and pass a block; the block
+    # will be passed the job object (which has a `ttl` method for getting the job's remaining TTL),
+    # and the block should return the desired timeout in seconds.
+    # This allows you to set a hard constant time out to a particular job class
+    # (using something like `extend Qless::Middleware::Timeout.new { 60 * 60 }`),
+    # or a variable timeout based on the individual TTLs of each job
+    # (using something like `extend Qless::Middleware::Timeout.new { |job| job.ttl * 1.1 }`).
+    class Timeout < Module
+      def initialize
+        module_eval do
+          define_method :around_perform do |job|
+            timeout_value = yield job
```

Comment:
I am not sure if this syntax is compatible with older versions of Ruby (Qless still supports 1.9.3). Please do not bother to comment if it does.